server/src/Icepeak/Server/WebsocketServer/HandleClientMultiSubscription.hs

```diff
@@ -0,0 +1,140 @@
+module Icepeak.Server.WebsocketServer.HandleClientMultiSubscription where
+
+import Control.Concurrent (modifyMVar_, newEmptyMVar)
+import Control.Concurrent.Async (withAsync)
+import Control.Concurrent.MVar (MVar, takeMVar, tryTakeMVar, newMVar, readMVar)
+import Control.Exception (SomeAsyncException, SomeException, catch, finally, fromException, throwIO)
+import Control.Monad (forever)
+import Data.Aeson (Value)
+import Data.Foldable (for_)
+import Data.Maybe (catMaybes)
+import Data.Text (Text)
+import Data.UUID (UUID)
+import System.Random (randomIO)
+
+import Data.HashMap.Strict (HashMap)
+import qualified Data.HashMap.Strict as HashMap
+
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString.Lazy as LBS
+import qualified Network.WebSockets as WS
+
+import Icepeak.Server.Core (Core, coreClients, withCoreMetrics, SubscriberState (SubscriberStateNew))
+import qualified Icepeak.Server.Metrics as Metrics
+import qualified Icepeak.Server.Subscription as Subscription
+
+-- * Client handling
+
+newUUID :: IO UUID
+newUUID = randomIO
+
+data ClientPayload
+  = Subscribe   { subscribePath :: [Text], token :: Text }
+  | UnSubscribe { unsubscribePath :: [Text] }
+
+instance Aeson.FromJSON ClientPayload where
+  parseJSON = undefined
+
+data ClientPayloadParsed
+  = ClientPayloadSuccessfulParse ClientPayload
+  | ClientPayloadFailedParse
+
+clientPayloadFromByteString :: LBS.ByteString -> ClientPayloadParsed
+clientPayloadFromByteString byteString = 
+  case Aeson.eitherDecode byteString of
+      (Left _jsonDecodeError) -> ClientPayloadFailedParse
+      (Right clientPayload)   -> ClientPayloadSuccessfulParse clientPayload
+
+clientPayloadFromDataMessage :: WS.DataMessage -> ClientPayloadParsed
+clientPayloadFromDataMessage (WS.Text encodedUtf8ByteString _mbDecodedCache) =
+  clientPayloadFromByteString encodedUtf8ByteString
+clientPayloadFromDataMessage (WS.Binary _) = ClientPayloadFailedParse
+  
+authorisePathSubscription :: [Text] -> Text -> IO Bool
+authorisePathSubscription = undefined
+
+handleClient :: WS.Connection -> Core -> IO ()
+handleClient conn core = do
+  uuid <- newUUID
+
+  isDirtyMVar <- newMVar ()
+  subscribedPathsMVar <- newMVar (HashMap.empty :: HashMap [Text] (MVar Value))
+
+  let
+    serverStateMVar = coreClients core
+
+    onConnect = withCoreMetrics core Metrics.incrementSubscribers
+
+    onDisconnect = do
+      paths <-  HashMap.keys <$> takeMVar subscribedPathsMVar
+      for_ paths
+        (\path -> modifyMVar_ serverStateMVar
+          (pure . Subscription.unsubscribe path uuid))
```

Comment:
We might be able to improve this setup some more. Instead of passing
`isDirtyMVar` into the tree, we could insert the actions instead. Exposing an
interface instead of the datatype. From what I can see, it would need an action
that 'dirties' it, along with actions that manipulate `subscribedPathsMVar`. WDYT?

Reply:
[52b5401](https://github.com/channable/icepeak/pull/119/commits/52b5401aee9f23afdf1a4e8cc7d56e3dbc03ac3f) Though I may want to refactor it to not need to pass a callback.